operator/src/main/java/oracle/kubernetes/operator/utils/ValidationUtils.java

```diff
@@ -0,0 +1,137 @@
+ * Domain or Cluster resource. It is used by the validating webhook.
+ *
+ * <p>The current version supports validation of the following:
+ * <ul>
+ * <li>The proposed replicas settings at the domain level and/or cluster level can be honored by WebLogic domain config.
+ * </li>
+ * </ul>
+ * </p>
+ */
+
+public class ValidationUtils {
+  private static final LoggingFacade LOGGER = LoggingFactory.getLogger("Webhook", "Operator");
+
+  private ValidationUtils() {
+  }
+
+  /**
+   * Validating a proposed domain resource against an existing domain resource. It returns true if the proposed changes
+   * in the proposed domain resource can be honored, otherwise, returns false.
+   *
+   * @param existingDomain domain that needs to be validated against.
+   * @param proposedDomain domain that needs to be validated.
+   * @return true if valid, otherwise false
+   */
+  public static boolean validateDomain(@NotNull Domain existingDomain, @NotNull Domain proposedDomain) {
+    LOGGER.fine("Validating domain " + proposedDomain + " against " + existingDomain);
+    return unchanged(existingDomain, proposedDomain)
+        || validReplicas(proposedDomain)
+        || shouldIntrospect(existingDomain, proposedDomain);
+  }
+
+  private static boolean unchanged(@NotNull Domain existingDomain, @NotNull Domain proposedDomain) {
+    return existingDomain == proposedDomain || specEquals(existingDomain, proposedDomain);
+  }
+
+  private static boolean specEquals(@NotNull Domain existingDomain, @NotNull Domain proposedDomain) {
+    return Optional.of(existingDomain)
+        .map(Domain::getSpec)
+        .map(s -> isProposedSpecUnchanged(s, proposedDomain))
+        .orElse(false);
+  }
+
+  private static boolean isProposedSpecUnchanged(@NotNull DomainSpec existingSpec, @NotNull Domain proposedDomain) {
+    return existingSpec == proposedDomain.getSpec() || Objects.equals(existingSpec, proposedDomain.getSpec());
+  }
+
+  private static boolean shouldIntrospect(@NotNull Domain existingDomain, @NotNull Domain proposedDomain) {
+    return !Objects.equals(existingDomain.getIntrospectVersion(), proposedDomain.getIntrospectVersion())
+        || imagesChanged(existingDomain, proposedDomain);
+  }
+
+  private static boolean imagesChanged(@NotNull Domain existingDomain, @NotNull Domain proposedDomain) {
+    if (!Objects.equals(existingDomain.getDomainHomeSourceType(), proposedDomain.getDomainHomeSourceType())) {
+      return true;
+    }
+    switch (proposedDomain.getDomainHomeSourceType()) {
+      case IMAGE:
+        return !Objects.equals(getImage(existingDomain), getImage(proposedDomain));
+      case FROM_MODEL:
+        if (noAuxiliaryImagesConfigured(existingDomain) && noAuxiliaryImagesConfigured(proposedDomain)) {
+          return !Objects.equals(getImage(existingDomain), getImage(proposedDomain));
+        } else {
+          return noAuxiliaryImagesConfigured(existingDomain)
```

Comment:
This block could probably be extracted into its own method, which could be given a descriptive name to explain what is going on, such as `areAuxiliaryImagesChanged`. That would make the top-level method more clear.